File: app/pricing/status.py
"""Green / Yellow / Red pre-check status (spec §4–§5)."""

from app.pricing.assumptions import CP_THRESHOLDS, GIFTING_ENTRY_MACHINE_HOURS
from app.pricing.models import Brand, DesignStatus, DesignStatusResult, SlicerMetrics


def evaluate_status(
    design_cp: float,
    target_sp: float,
    brand: Brand,
    metrics: SlicerMetrics | None = None,
    thresholds: tuple[float, float] | None = None,
    entry_machine_hours: float | None = None,
    entry_rung: int = 999,
) -> DesignStatusResult:
    """Classify a design against the CP-share and machine-time rules.

    `thresholds`, `entry_machine_hours` and `entry_rung` default to the brand's
    built-in policy so the engine works standalone; the API passes the
    admin-editable values loaded from the brand record.
    """
    if target_sp <= 0:
        raise ValueError("target_sp must be positive")

    if thresholds is None:
        thresholds = CP_THRESHOLDS[brand]
    green_max, yellow_max = thresholds
    cp_pct = round(design_cp / target_sp, 4)
    reasons: list[str] = []

    if cp_pct <= green_max:
        status = DesignStatus.GREEN
    elif cp_pct <= yellow_max:
        status = DesignStatus.YELLOW
        reasons.append(f"Design CP is {cp_pct:.0%} of price — above the {green_max:.0%} target.")
    else:
        status = DesignStatus.RED
        reasons.append(f"Design CP is {cp_pct:.0%} of price — over the {yellow_max:.0%} ceiling.")

    # Entry-tier machine-time rule (spec §4): gifting must batch under the target.
    # When no explicit target is given, fall back to gifting's built-in policy so
    # standalone behaviour is unchanged; decor has no such rule.
    if entry_machine_hours is None and brand == Brand.GIFTING:
        entry_machine_hours = GIFTING_ENTRY_MACHINE_HOURS
        entry_rung = 999
    if (
        entry_machine_hours is not None
        and target_sp <= entry_rung
        and metrics is not None
        and metrics.effective_machine_time_hr > entry_machine_hours
    ):
        reasons.append(
            f"Effective machine time {metrics.effective_machine_time_hr:.2f}h exceeds the "
            f"{entry_machine_hours:.0f}h target for ₹{entry_rung}."
        )
        if status == DesignStatus.GREEN:
            status = DesignStatus.YELLOW

    return DesignStatusResult(status=status, cp_pct=cp_pct, reasons=reasons)

File: internal/pricing/status.go
package pricing

import "fmt"

// EvaluateStatus produces the Green / Yellow / Red verdict from the Design CP as
// a share of the target selling price, plus the entry-tier machine-time rule.
//
// thresholds is an optional (green_max, yellow_max) pair; when nil the brand's
// built-in thresholds are used. entryMachineHours is optional; entryRung is the
// price the machine-time rule applies at or below (the API passes 999 when the
// brand has no policy). It mirrors evaluate_status in app/pricing/status.py.
func EvaluateStatus(
	designCP, targetSP float64,
	brand Brand,
	metrics *SlicerMetrics,
	thresholds *[2]float64,
	entryMachineHours *float64,
	entryRung int,
) (DesignStatusResult, error) {
	if targetSP <= 0 {
		return DesignStatusResult{}, fmt.Errorf("target selling price must be positive")
	}

	var greenMax, yellowMax float64
	if thresholds != nil {
		greenMax, yellowMax = thresholds[0], thresholds[1]
	} else {
		greenMax, yellowMax = CPThresholds(brand)
	}

	cpPct := roundHalfEven(designCP/targetSP, 4)
	reasons := []string{}

	var status DesignStatus
	switch {
	case cpPct <= greenMax:
		status = StatusGreen
	case cpPct <= yellowMax:
		status = StatusYellow
		reasons = append(reasons, fmt.Sprintf(
			"Design CP is %s of price — above the %s target.",
			formatPercent0(cpPct), formatPercent0(greenMax),
		))
	default:
		status = StatusRed
		reasons = append(reasons, fmt.Sprintf(
			"Design CP is %s of price — over the %s ceiling.",
			formatPercent0(cpPct), formatPercent0(yellowMax),
		))
	}

	// Entry-tier machine-time rule. Gifting with no explicit policy falls back to
	// the built-in (2.0h at ₹999). The rule only ever downgrades GREEN to YELLOW,
	// but it always appends its reason when triggered.
	emh := entryMachineHours
	er := entryRung
	if emh == nil && brand == BrandGifting {
		v := GiftingEntryMachineHours
		emh = &v
		er = GiftingEntryRung
	}

	if emh != nil && targetSP <= float64(er) &&
		metrics != nil && metrics.EffectiveMachineTimeHr > *emh {
		reasons = append(reasons, fmt.Sprintf(
			"Effective machine time %sh exceeds the %sh target for ₹%d.",
			formatFixed(metrics.EffectiveMachineTimeHr, 2), formatFixed(*emh, 0), er,
		))
		if status == StatusGreen {
			status = StatusYellow
		}
	}

	return DesignStatusResult{Status: status, CPPct: cpPct, Reasons: reasons}, nil
}
